crypto/keys/ed25519/ed25519.go

```diff
@@ -0,0 +1,232 @@
+package ed25519
+
+import (
+	"crypto/subtle"
+	"fmt"
+	"io"
+
+	"github.com/tendermint/tendermint/crypto"
+	tmed25519 "github.com/tendermint/tendermint/crypto/ed25519"
+	"github.com/tendermint/tendermint/crypto/tmhash"
+	"golang.org/x/crypto/ed25519"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	cryptotypes "github.com/cosmos/cosmos-sdk/crypto/types"
+)
+
+//-------------------------------------
+
+const (
+	PrivKeyName = "cosmos/PrivKeyEd25519"
+	PubKeyName  = "cosmos/PubKeyEd25519"
+	// PubKeySize is is the size, in bytes, of public keys as used in this package.
+	PubKeySize = 32
+	// PrivateKeySize is the size, in bytes, of private keys as used in this package.
+	PrivateKeySize = 64
+	// Size of an Edwards25519 signature. Namely the size of a compressed
+	// Edwards25519 point, and a field element. Both of which are 32 bytes.
+	SignatureSize = 64
+	// SeedSize is the size, in bytes, of private key seeds. These are the
+	// private key representations used by RFC 8032.
+	SeedSize = 32
+
+	keyType = "ed25519"
+)
+
+var _ cryptotypes.PrivKey = &PrivKey{}
+var _ codec.AminoMarshaler = &PrivKey{}
+
+// Bytes returns the privkey byte format.
+func (privKey *PrivKey) Bytes() []byte {
+	return privKey.Key
+}
+
+// Sign produces a signature on the provided message.
+// This assumes the privkey is wellformed in the golang format.
+// The first 32 bytes should be random,
+// corresponding to the normal ed25519 private key.
+// The latter 32 bytes should be the compressed public key.
+// If these conditions aren't met, Sign will panic or produce an
+// incorrect signature.
+func (privKey *PrivKey) Sign(msg []byte) ([]byte, error) {
+	signatureBytes := ed25519.Sign(ed25519.PrivateKey(privKey.Key), msg)
+	return signatureBytes, nil
+}
+
+// PubKey gets the corresponding public key from the private key.
+//
+// Panics if the private key is not initialized.
+func (privKey *PrivKey) PubKey() crypto.PubKey {
+	// If the latter 32 bytes of the privkey are all zero, privkey is not
+	// initialized.
+	initialized := false
+	for _, v := range privKey.Key[32:] {
+		if v != 0 {
+			initialized = true
+			break
+		}
+	}
+
+	if !initialized {
+		panic("Expected ed25519 PrivKey to include concatenated pubkey bytes")
+	}
+
+	pubkeyBytes := make([]byte, PubKeySize)
+	copy(pubkeyBytes, privKey.Key[32:])
+	return &PubKey{Key: pubkeyBytes}
```

Comment:
1. Why do we combine public key in private key?
2. ed22519 public key is derived (1-1) from a private key, so even if we have only private key we can make a public key

Reply:
this is directly adapted from https://github.com/tendermint/tendermint/blob/master/crypto/ed25519/ed25519.go#L65

Reply:
OK, but why do we need it in SDK?

Reply:
so that we have our own implementation of ed25519